README.md
[![Build Status](https://travis-ci.org/mapbox/tilelive-overlay.svg?branch=master)](https://travis-ci.org/mapbox/tilelive-overlay)

# tilelive-overlay

Add GeoJSON features with simplestyle styles in a tilelive pipeline.

## Install

    npm install --save tilelive-overlay

## API

* `overlaydata://GEOJSONSTRING`
* `overlaydata://2x:GEOJSONSTRING` (retina)

Provides a `overlaydata://GEOJSONSTRING` scheme for [tilelive](https://www.npmjs.org/package/tilelive),
which transforms [GeoJSON](http://geojson.org/) into a [Mapnik](http://mapnik.org/) stylesheet
using [geojson-mapnikify](https://github.com/mapbox/geojson-mapnikify), and then provides
raster tiles of it.


CHANGELOG.md
# Changelog

## v2.3.0
- Avoids potential URL parsing issue and instead just does a simple check for the overlaydata protocol
- Reverts back to @mapbox/geojson-mapnikify 2.1.0

## v2.2.0
- Updated to @mapbox/geojson-mapnikify 3.0.0 (node 10 support)
- Updated to @mapbox/sphericalmercator 1.1.0 (support for floating point zoom levels in `ll` and `px` methods)

## v2.1.0
- Updated to @mapbox/geojson-mapnikify 2.1.0 (with improved error messaging)

## v2.0.0
- Updated to @mapbox/geojson-mapnikify 2.0.0
- Now accepting either mapnik 3.x or 4.x

## v1.1.0
- Updated version of node-mapnik to 4.0.1
- Ends node v0.10.x support

## v1.0.0

 - Updated to use mapnik 3.7.0
 - Updated geojson-mapnikify to 1.0.0
 - Dropped support for windows

## v0.8.0

 - Updated to use mapnik 3.6.0
 - Modified dependencies to use mapbox NPM namespace


test/marker.js
var test = require('tap').test,
    fs = require('fs'),
    getUrlMarker = require('../lib/urlmarker.js');

test('getUrlMarker', function(t) {
    getUrlMarker({
        label: 'https://www.mapbox.com/maki/renders/bus-18.png'
    }, function(err, data) {
        t.notOk(err, 'does not return an error');
        t.ok(data, 'returns data');
        t.end();
    });
});

test('getUrlMarker-invalid', function(t) {
    getUrlMarker({
        label: 'https://wwwmapbox.com/maki/renders/bus-18.png'
    }, function(err, data) {
        t.equal(err.message, 'Unable to load marker from URL.');
        t.notOk(data, 'does not return data');
        t.end();
    });
});

test('getUrlMarker-invalid', function(t) {
    getUrlMarker({
        label: 'https//wwwmapbox.com/maki/renders/bus-18.png'
    }, function(err, data) {
        t.equal(err.message, 'Unable to load marker from URL.');
        t.notOk(data, 'does not return data');
        t.end();
    });
});

test('getUrlMarker-invalid', function(t) {
    getUrlMarker({
        label: 'https/wwwmapbox.com/maki/renders/bus-18.png'
    }, function(err, data) {
        t.equal(err.message, 'Unable to load marker from URL.');
        t.notOk(data, 'does not return data');
        t.end();
    });
});


test/simple.js
var test = require('tap').test,
    Simple = require('../');

test('simple', function(t) {
    new Simple('simple://' + __dirname + '/data/example.geojson', function(err, source) {
        if (err) throw err;
        t.notOk(err, 'no error returned');
        t.ok(source, 'source created');
        source.getGrid(0, 0, 0, function(err, res) {
            t.equal(err, 'This source does not provide grids', 'does not provide grids');
        });
        source.getInfo(function(err, res) {
            t.equal(err, 'This source does not provide info', 'does not provide info');
        });
        t.end();
    });
});


test/overlay.js
var test = require('tap').test,
    fs = require('fs'),
    mapnik = require('@kartotherian/mapnik'),
    Overlay = require('../');

test('overlay', function(t) {
    new Overlay('overlaydata://' + fs.readFileSync(__dirname + '/data/example.geojson', 'utf8'),
        function(err, source) {
        if (err) throw err;
        t.notOk(err, 'no error returned');
        t.ok(source, 'source created');
        source.getGrid(0, 0, 0, function(err, res) {
            t.equal(err, 'This source does not provide grids', 'does not provide grids');
        });
        source.getInfo(function(err, res) {
            t.equal(err, 'This source does not provide info', 'does not provide info');
        });
        t.end();
    });
});

textFixture('example');
textFixture('example_defaults');
textFixture('complex');
textFixture('line');
textFixture('polygon');
textFixture('polygon_style');
textFixture('polygon_nofill');
textFixture('polygon_nostroke');

function compare(res, filepath, assert) {
    if (process.env.UPDATE) fs.writeFileSync(filepath, res);
    var expectImage = new mapnik.Image.open(filepath);
    var resultImage = new mapnik.Image.fromBytesSync(res);

    // Allow < 2% of pixels to vary by > default comparison threshold of 16.
    var pxThresh = resultImage.width() * resultImage.height() * 0.005;
    var pxDiff = expectImage.compare(resultImage);
    assert.ok(pxDiff < pxThresh, filepath);
}

function textFixture(name) {
    test('overlay - ' + name, function(t) {
        new Overlay('overlaydata://' + fs.readFileSync(__dirname + '/data/' + name + '.geojson', 'utf8'),
            function(err, source) {
            t.notOk(err, 'no error returned');
            t.ok(source, 'source created');
            source.getTile(0, 0, 0, function(err, res) {
                t.notOk(err, 'no error');
                compare(res, __dirname + '/data/' + name + '.png', t);
                t.end();
            });
        });
    });
    test('overlay - ' + name + ' @2x', function(t) {
        new Overlay('overlaydata://2x:' + fs.readFileSync(__dirname + '/data/' + name + '.geojson', 'utf8'),
            function(err, source) {
            t.notOk(err, 'no error returned');
            t.ok(source, 'source created');
            source.getTile(0, 0, 0, function(err, res) {
                t.notOk(err, 'no error');
                compare(res, __dirname + '/data/' + name + '@2x.png', t);
                t.end();
            });
        });
    });
    test('overlay - ' + name + ' @2x (legacy)', function(t) {
        new Overlay('overlaydata://2x:legacy:' + fs.readFileSync(__dirname + '/data/' + name + '.geojson', 'utf8'),
            function(err, source) {
            t.notOk(err, 'no error returned');
            t.ok(source, 'source created');
            source.getTile(0, 0, 0, function(err, res) {
                t.notOk(err, 'no error');
                compare(res, __dirname + '/data/' + name + '@2x.legacy.png', t);
                t.end();
            });
        });
    });
}

test('overlay buffer', function(t) {
    var source;
    new Overlay('overlaydata://' + fs.readFileSync(__dirname + '/data/buffer.geojson', 'utf8'),
        function(err, s) {
        if (err) throw err;
        source = s;
        t.notOk(err, 'no error returned');
        t.ok(source, 'source created');
        a();
    });
    function a() {
        source.getTile(1, 0, 0, function(err, res) {
            t.notOk(err, 'no error');
            compare(res, __dirname + '/data/buffer-1.0.0.png', t);
            b();
        });
    }
    function b() {
        source.getTile(1, 1, 0, function(err, res) {
            t.notOk(err, 'no error');
            compare(res, __dirname + '/data/buffer-1.1.0.png', t);
            t.end();
        });
    }
});

test('overlay buffer @2x', function(t) {
    var source;
    new Overlay('overlaydata://2x:' + fs.readFileSync(__dirname + '/data/buffer.geojson', 'utf8'),
        function(err, s) {
        if (err) throw err;
        source = s;
        t.notOk(err, 'no error returned');
        t.ok(source, 'source created');
        a();
    });
    function a() {
        source.getTile(1, 0, 0, function(err, res) {
            t.notOk(err, 'no error');
            compare(res, __dirname + '/data/buffer-1.0.0@2x.png', t);
            b();
        });
    }
    function b() {
        source.getTile(1, 1, 0, function(err, res) {
            t.notOk(err, 'no error');
            compare(res, __dirname + '/data/buffer-1.1.0@2x.png', t);
            t.end();
        });
    }
});


test('overlay-invalid', function(t) {
    new Overlay('overlaydata://invalidjson', function(err, source) {
        t.equal(err, 'invalid geojson');
        t.end();
    });
});

test('overlay-invalid', function(t) {
    new Overlay('overladata://invalidjson', function(err, source) {
        t.equal(err, 'Only the overlaydata protocol is supported');
        t.end();
    });
});


lib/get.js
var xtend = require('xtend'),
    url = require('url'),
    ErrorHTTP = require('./errorhttp'),
    request = require('request'),
    agent = new (require('agentkeepalive'))({
        maxSockets: 128,
        maxKeepAliveRequests: 0,
        maxKeepAliveTime: 30000
    }),
    urlcache = require('lru-cache')({ max: 100, maxAge: 600000 });

module.exports = get;

/**
 * @param {string} uri
 * @param {function} callback
 */
function get(uri, key, callback) {

    if (urlcache.get(key)) {
        return callback(null, urlcache.get(key));
    }

    // This might load the same file multiple times, but the overhead should
    // be very little.
    request({
        uri: uri,
        timeout: 5000,
        headers: { 'accept-encoding': 'binary' },
        encoding: 'binary',
        // Don't use keepalive agent for https.
        agent: uri.indexOf('https') === 0 ? null : agent
    }, function(err, resp, data) {
        // Use err.status of 400 as it's not an unexpected application error,
        // but likely due to a bad request. Catches ECONNREFUSED,
        // getaddrinfo ENOENT, etc.
        if (err || !data || !resp || resp.statusCode != 200) {
            return callback(new ErrorHTTP('Unable to load marker from URL.', 400));
        }

        // request 2.2.x *always* returns the response body as a string.
        // @TODO remove this once request is upgraded.
        if (!(data instanceof Buffer)) data = new Buffer(data, 'binary');

        // Restrict data length.
        if (data.length > 32768) return callback(new ErrorHTTP(400));

        var headers = resp.headers;

        urlcache.set(key, data);
        callback(null, data);
    });
}


lib/marker.js
var fs = require('fs'),
    path = require('path'),
    blend = require('blend'),
    xtend = require('xtend'),
    get = require('./get.js');

var ErrorHTTP = require('./errorhttp');

// Hardcoded mtime date to ensure stable Last-Modified / ETag headers for
// marker requests. This should be updated whenever the source marker assets
// or rendering behavior is changed.
// @TODO deprecate this in favor of loading markers from S3/API source.
var mtime = new Date('Mon May 06 2013 20:00:00 UTC');

module.exports = Marker;

function Marker(options, callback) {
    if (options.tint) {
        // Expand hex shorthand (3 chars) to 6, e.g. 333 => 333333.
        // This is not done upstream in `node-tint` as some such
        // shorthand cannot be disambiguated from other tintspec strings,
        // e.g. 123 (rgb shorthand) vs. 123 (hue).
        if (options.tint.length === 3) options.tint =
            options.tint[0] + options.tint[0] +
            options.tint[1] + options.tint[1] +
            options.tint[2] + options.tint[2];
        this.tint = blend.parseTintString(options.tint);
    }

    // Unique key for all (file or url) markers.
    this.key = options.name + (typeof options.label !== 'undefined' ? '-' + options.label : '') + (options.retina ? '@2x' : '');

    if (options.name === 'url') {
        this._loadURL(options, callback);
    // Alpha-numeric files.
    } else if (!options.label || (options.label && options.label.length === 1)) {
        this._loadFile(options, callback);
    } else {
        this._loadMaki(options, callback);
    }
}

Marker.offsets = {
    's': {x:4,y:4},
    'm': {x:6,y:5},
    'l': {x:5,y:7},
    's@2x': {x:8,y:8},
    'm@2x': {x:12,y:10},
    'l@2x': {x:10,y:14}
};

Marker.sizes = {
    s: 12,
    m: 18,
    l: 24
};

// Loads up all default markers at require time.
Marker.cache = [
    ['base', 'base'],
    ['mask', 'mask'],
    ['alphanum', 'symbol']
].reduce(function(memo, destsrc) {
    var dest = destsrc[1],
        src = destsrc[0];
    var basepath = path.resolve(__dirname + '/../markers-src/' + src);
    memo[dest] = fs.readdirSync(basepath)
        .sort()
        .reduce(function(memo, file) {
            if (path.extname(file) !== '.png') return memo;
            var key = path.basename(file, '.png')
                .replace('-12', '-s')
                .replace('-18', '-m')
                .replace('-24', '-l');
            memo[key] = fs.readFileSync(basepath + '/' + file);
            return memo;
        }, memo[dest] || {});
    return memo;
}, { url: {} });

/**
 * @param {object} options
 * @param {function} callback
 */
Marker.prototype._loadMaki = function(options, callback) {
    var base = options.name + ((options.name && options.retina) ? '@2x' : '');
    var size = options.name.split('-').pop();
    var symbol = (options.label || '') +
        ((options.label && size) ? '-' + Marker.sizes[size] : '') +
        ((options.label && options.retina) ? '@2x' : '');

    if (!base || !size) {
        return callback(new ErrorHTTP('Marker "' + this.key + '" is invalid.', 400));
    }

    get('http://s3.amazonaws.com/www.mapbox.com/maki/renders/' + symbol + '.png', this.key, function(err, data) {
        if (err) return callback(new ErrorHTTP('Marker "' + this.key + '" is invalid.', 400));

        // Base marker gets tint applied.
        var parts = [{
            buffer: Marker.cache.base[base],
            tint: this.tint
        }];

        // If symbol is present, find correct offset (varies by marker size).
        if (symbol) {
            parts.push(xtend({
                buffer: data,
                tint: blend.parseTintString('0x0;0x0;1.4x0'),
            }, Marker.offsets[size + (options.retina ? '@2x' : '')]));
        }

        // Add mask layer.
        parts.push({
            buffer: Marker.cache.mask[base]
        });

        // Extract width and height from the IHDR. The IHDR chunk must appear
        // first, so the location is always fixed.
        this.width = Marker.cache.base[base].readUInt32BE(16);
        this.height = Marker.cache.base[base].readUInt32BE(20);

        this.maxAge = 3600;
        this.mtime = mtime;
        this.size = data.length;

        // Combine base, (optional) symbol, to supply the final marker.
        blend(parts, {
            format: 'png',
            quality: 256,
            width: this.width,
            height: this.height
        }, function(err, data) {
            if (err) return callback(err);
            this.image = data;
            return callback(null, this);
        }.bind(this));
    }.bind(this));
};

/**
 * @param {object} options
 * @param {function} callback
 */
Marker.prototype._loadURL = function(options, callback) {
    var uri = options.label;
    if (uri.substring(0, 4) !== 'http') uri = 'http://' + uri;

    get(uri, function(err, data) {
        if (err) return callback(err);

        // Check for PNG header.
        if (data.toString('binary', 0, 8) !== '\x89\x50\x4E\x47\x0D\x0A\x1A\x0A') {
            return callback(new ErrorHTTP('Marker image format is not supported.', 415));
        }

        // Extract width and height from the IHDR. The IHDR chunk must appear
        // first, so the location is always fixed.
        this.width = data.readUInt32BE(16);
        this.height = data.readUInt32BE(20);

        // Check image size. 400x400 square limit.
        if (this.width * this.height > 160000) {
            return callback(new ErrorHTTP('Marker image size must not exceed 160000 pixels.', 415));
        }

        this.maxAge = 3600;
        this.mtime = mtime;
        this.size = data.length;
        this.image = data;

        if (!this.tint) return callback(null, this);

        blend([{buffer:data, tint:this.tint}], {}, function(err, tinted) {
            if (err) return callback(err);
            this.image = tinted;
            return callback(null, this);
        }.bind(this));
    }.bind(this));
};

/**
 * @param {object} options
 * @param {function} callback
 */
Marker.prototype._loadFile = function(options, callback) {
    var base = options.name + ((options.name && options.retina) ? '@2x' : '');
    var size = options.name.split('-').pop();
    var symbol = (options.label || '') +
        ((options.label && size) ? '-' + size : '') +
        ((options.label && options.retina) ? '@2x' : '');

    if (!base || !size || !Marker.cache.base[base] || (symbol && !Marker.cache.symbol[symbol])) {
        return callback(new ErrorHTTP('Marker "' + this.key + '" is invalid.', 400));
    }

    // Base marker gets tint applied.
    var parts = [{
        buffer: Marker.cache.base[base],
        tint: this.tint
    }];

    // If symbol is present, find correct offset (varies by marker size).
    if (symbol) {
        parts.push(xtend({
            buffer: Marker.cache.symbol[symbol],
            tint: blend.parseTintString('0x0;0x0;1.4x0'),
        }, Marker.offsets[size + (options.retina ? '@2x' : '')]));
    }

    // Add mask layer.
    parts.push({ buffer:Marker.cache.mask[base] });

    // Extract width and height from the IHDR. The IHDR chunk must appear
    // first, so the location is always fixed.
    this.width = Marker.cache.base[base].readUInt32BE(16);
    this.height = Marker.cache.base[base].readUInt32BE(20);
    this.maxAge = 3600;

    // Combine base, (optional) symbol, to supply the final marker.
    blend(parts, {
        format: 'png',
        quality: 256,
        width: this.width,
        height: this.height
    }, function(err, data) {
        if (err) return callback(err);
        this.mtime = mtime;
        this.size = data.length;
        this.image = data;
        return callback(null, this);
    }.bind(this));
};


lib/urlmarker.js
var get = require('./get');

/**
 * Load an image from a URL for use as a custom marker.
 *
 * @param {object} options
 * @param {function} callback
 */
module.exports = function(options, callback) {
    var uri = options.label;

    if (options.tint) {
        // Expand hex shorthand (3 chars) to 6, e.g. 333 => 333333.
        // This is not done upstream in `node-tint` as some such
        // shorthand cannot be disambiguated from other tintspec strings,
        // e.g. 123 (rgb shorthand) vs. 123 (hue).
        if (options.tint.length === 3) options.tint =
            options.tint[0] + options.tint[0] +
            options.tint[1] + options.tint[1] +
            options.tint[2] + options.tint[2];
        options.parsedTint = blend.parseTintString(options.tint);
    }

    if (uri.substring(0, 4) !== 'http') uri = 'http://' + uri;

    get(uri, uri, function(err, data) {
        if (err) return callback(err);

        // Check for PNG header.
        if (data.toString('binary', 0, 8) !== '\x89\x50\x4E\x47\x0D\x0A\x1A\x0A') {
            return callback(new ErrorHTTP('Marker image format is not supported.', 415));
        }

        // Extract width and height from the IHDR. The IHDR chunk must appear
        // first, so the location is always fixed.
        var width = data.readUInt32BE(16),
            height = data.readUInt32BE(20);

        // Check image size. 400x400 square limit.
        if (width * height > 160000) {
            return callback(new ErrorHTTP('Marker image size must not exceed 160000 pixels.', 415));
        }

        if (!options.parsedTint) return callback(null, {
            image: data,
            size: data.length,
            width: width,
            height: height
        });

        blend([{
            buffer:data,
            tint: options.parsedTint
        }], {}, function(err, tinted) {
            if (err) return callback(err);
            return callback(null, {
                image: tinted,
                size: data.length,
                width: width,
                height: height
            });
        });
    });
};


benchmark/bench.js
var Benchmark = require('benchmark'),
    fs = require('fs'),
    queue = require('queue-async'),
    Overlay = require('../');

var suite = new Benchmark.Suite();

new Overlay('overlaydata://' + fs.readFileSync('./test/data/complex.geojson', 'utf8'),
    function(err, source) {
        if (err) throw err;
        suite
        .add('#getTile', function(deferred) {
            var q = queue(4);
            for (var i = 0; i < 16; i++) q.defer(function(done) {
                source.getTile(11, 486, 779, function(err,im) {
                    if (err) throw err;
                    if (im.length != 4605) {
                        throw new Error("image is not the length we expected: " + im.length);
                    }
                    done();
                });
            });
            q.awaitAll(function() {
                deferred.resolve();
            });
        }, {
            defer: true
        })
        .on('cycle', function(event) {
            console.log(String(event.target));
        })
        .on('complete', function() {
            console.log('Fastest is ' + this.filter('fastest').pluck('name'));
        })
        .run({ 'async': true });
    });


lib/generatexml.js
var normalize = require('geojson-normalize'),
    PROJ = "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over",
    WGS84 = "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs",
    HEADER = '<?xml version="1.0" encoding="utf-8"?>' +
    '<Map srs="' + PROJ + '">',
    FOOTER = '</Map>',
    xtend = require('xtend');

var styleMap = {
    'stroke': ['LineSymbolizer', 'stroke'],
    'stroke-opacity': ['LineSymbolizer', 'stroke-opacity'],
    'stroke-width': ['LineSymbolizer', 'stroke-width'],
    'fill': ['PolygonSymbolizer', 'fill'],
    'fill-opacity': ['PolygonSymbolizer', 'fill-opacity']
};

var defaultFilled = {
    fill: '#555555',
    'fill-opacity': 0.6
};

var defaultStroked = {
    stroke: '#555555',
    'stroke-width': 2,
    'stroke-opacity': 1,
};

var typedDefaults = {
    LineString: defaultStroked,
    MultiLineString: defaultStroked,
    Polygon: defaultFilled,
    MultiPolygon: defaultFilled
};

/**
 * @param {object} data a geojson object
 * @returns {string} a mapnik style
 */
module.exports = function generateXML(data, TMP, retina) {
    var gj = normalize(data);
    if (!gj) return null;
    var ls = gj.features.map(convertFeature(TMP, retina));

    return {
        xml: HEADER +
            ls.map(function(_) { return _.style; }).join('') +
            ls.map(function(_) { return _.layer; }).join('') +
            FOOTER,
        resources: ls.reduce(function(mem, _) {
            return mem.concat(_.resources);
        }, [])
    };
};

/**
 * @param {object} feature
 * @returns {string}
 */
function markerString(feature, retina) {
    var fp = feature.properties || {},
        size = fp['marker-size'] || 'medium',
        symbol = (fp['marker-symbol']) ? '-' + fp['marker-symbol'] : '',
        color = (fp['marker-color'] || '7e7e7e').replace('#', '');

    return 'pin-' + size.charAt(0) + symbol + '+' + color + (retina ? '@2x' : '');
}

/**
 * @param {object} feature
 * @returns {string}
 */
function markerURL(feature) {
    return (feature.properties || {})['marker-url'];
}

/**
 * @param {object} feature geojson feature
 * @returns {object}
 */
function convertFeature(TMP, retina) {
    return function(feature, i) {
        var style = generateStyle(feature, i, TMP, retina);
        return {
            style: style.style,
            resources: style.resources,
            layer: generateLayer(feature, i)
        };
    };
}

/**
 * @param {object} a
 * @param {object} b
 * @returns {object}
 */
function fallback(a, b) {
    var c = {};
    for (var k in b) {
        if (a[k] === undefined) c[k] = b[k];
        else c[k] = a[k];
    }
    return c;
}

/**
 * @param {object} feature geojson feature
 * @returns {string}
 */
function generateStyle(feature, i, TMP, retina) {
    var defaults = typedDefaults[feature.geometry.type] || {},
        props = pairs(xtend({}, defaults, feature.properties || {})),
        symbolizerGroups = props.reduce(collectSymbolizers, {}),
        markerStyle = '',
        resources = [];

    if (feature.geometry.type === 'Point' ||
        feature.geometry.type === 'MultiPoint') {
        if (markerURL(feature)) {
            resources.push(markerURL(feature));
            markerStyle = tagClose('PointSymbolizer', [
                ['file', TMP + markerString(feature)]
            ]);
        } else {
            resources.push(markerString(feature, retina));
            markerStyle = tagClose('PointSymbolizer', [
                ['file', TMP + '/' + markerString(feature, retina) + '.png']
            ]);
        }
    }

    return {
        style: tag('Style',
            tag('Rule',
            pairs(symbolizerGroups)
                .map(function(symbolizer) {
                    return tagClose(symbolizer[0], pairs(symbolizer[1]));
                }).join('') + markerStyle),
                [['name', 'style-' + i]]),
        resources: resources
    };
}

/**
 * @param {object} mem
 * @param {array} prop
 * @returns {object}
 */
function collectSymbolizers(mem, prop) {
    var mapped = styleMap[prop[0]];
    if (mapped) {
        if (!mem[mapped[0]]) mem[mapped[0]] = {};
        mem[mapped[0]][mapped[1]] = prop[1];
    }
    return mem;
}

/**
 * @param {object} feature geojson feature
 * @returns {string}
 */
function generateLayer(feature, i) {
    if (!feature.geometry) return null;

    return tag('Layer',
        tag('StyleName', 'style-' + i) +
        tag('Datasource',
            [
                ['type', 'ogr'],
                ['layer_by_index', '0'],
                ['driver', 'GeoJson'],
                ['string', JSON.stringify(feature.geometry)]
            ].map(function(a) {
                return tag('Parameter', a[1], [['name', a[0]]]);
            }).join('')), [
                ['name', 'layer-' + i],
                ['srs', WGS84]
            ]);
}

/**
 * @param {object} o
 * @returns {array}
 */
function pairs(o) {
    return Object.keys(o).map(function(k) {
        return [k, o[k]];
    });
}

/**
 * @param {array} _ an array of attributes
 * @returns {string}
 */
function attr(_) {
    return (_ && _.length) ? (' ' + _.map(function(a) {
        return a[0] + '="' + a[1] + '"';
    }).join(' ')) : '';
}

/**
 * @param {string} el element name
 * @param {array} attributes array of pairs
 * @returns {string}
 */
function tagClose(el, attributes) {
    return '<' + el + attr(attributes) + '/>';
}

/**
 * @param {string} el element name
 * @param {string} contents innerXML
 * @param {array} attributes array of pairs
 * @returns {string}
 */
function tag(el, contents, attributes) {
    return '<' + el + attr(attributes) + '>' + contents + '</' + el + '>';
}

/**
 * @param {string} _ a string of attribute
 * @returns {string}
 */
function encode(_) {
    return (_ === null ? '' : _.toString()).replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;');
}


lib/sphericalmercator.js
var mapnik = require('mapnik');

var proj4 = '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over';
var mercator = new mapnik.Projection(proj4);

/**
 * SphericalMercator constructor: precaches calculations
 * for fast tile lookups
 */
function SphericalMercator() {
    var size = 256;
    this.Bc = [];
    this.Cc = [];
    this.zc = [];
    this.Ac = [];
    this.DEG_TO_RAD = Math.PI / 180;
    this.RAD_TO_DEG = 180 / Math.PI;
    this.size = 256;
    this.levels = 40;
    this.proj4 = proj4;
    for (var d = 0; d < this.levels; d++) {
        this.Bc.push(size / 360);
        this.Cc.push(size / (2 * Math.PI));
        this.zc.push(size / 2);
        this.Ac.push(size);
        size *= 2;
    }
}

/**
 * Get the max of the first two numbers and the min of that and the third
 *
 * @param {Number} a the first number.
 * @param {Number} b the second number.
 * @param {Number} c the third number.
 * @return {Number}
 */
SphericalMercator.prototype.minmax = function(a, b, c) {
    return Math.min(Math.max(a, b), c);
};

/**
 * Convert lat lon to screen pixel value
 *
 * @param {Array} px [lat lon] array of geographic coordinates.
 * @param {Number} zoom number of the zoom level.
 */
SphericalMercator.prototype.ll_to_px = function(ll, zoom) {
    var d = this.zc[zoom];
    var f = this.minmax(Math.sin(this.DEG_TO_RAD * ll[1]), -0.9999, 0.9999);
    var x = Math.round(d + ll[0] * this.Bc[zoom]);
    var y = Math.round(d + 0.5 * Math.log((1 + f) / (1 - f)) * (-this.Cc[zoom]));
    return [x, y];
};

/**
 * Convert screen pixel value to lat lon
 *
 * @param {Array} px [x y] array of geographic coordinates.
 * @param {Number} zoom number of the zoom level.
 */
SphericalMercator.prototype.px_to_ll = function(px, zoom) {
    var zoom_denom = this.zc[zoom];
    var g = (px[1] - zoom_denom) / (-this.Cc[zoom]);
    var lat = (px[0] - zoom_denom) / this.Bc[zoom];
    var lon = this.RAD_TO_DEG * (2 * Math.atan(Math.exp(g)) - 0.5 * Math.PI);
    return [lat, lon];
};

/**
 * Convert tile xyz value to Mapnik envelope
 *
 * @param {Number} x latitude number.
 * @param {Number} y longitude number.
 * @param {Number} zoom zoom.
 * @param {Boolean} tms_style whether to compute a tms tile.
 * @return Object Mapnik envelope.
 */
SphericalMercator.prototype.xyz_to_envelope = function(x, y, zoom, TMS_SCHEME) {
    if (TMS_SCHEME) {
        y = (Math.pow(2, zoom) - 1) - y;
    }
    var ll = [x * this.size, (y + 1) * this.size];
    var ur = [(x + 1) * this.size, y * this.size];
    var bbox = this.px_to_ll(ll, zoom).concat(this.px_to_ll(ur, zoom));
    return mercator.forward(bbox);
};

module.exports = new SphericalMercator();



index.js
var mapnik = require('mapnik'),
    sm = new (require('@mapbox/sphericalmercator'))(),
    mapnikify = require('@mapbox/geojson-mapnikify');

if (mapnik.register_default_input_plugins) {
    mapnik.register_default_input_plugins();
}

module.exports = Source;

require('util').inherits(Source, require('events').EventEmitter);

/**
 * Create a new source that returns tiles from a simplestyle-supporting
 * GeoJSON object.
 *
 * @param {string} uri
 * @param {function} callback
 * @returns {undefined}
 */
function Source(id, callback) {
    if (!id.startsWith('overlaydata://')) {
        return callback('Only the overlaydata protocol is supported');
    }

    var data = id.replace('overlaydata://', '');
    var retina = false;
    var legacy = false;

    if (data.indexOf('2x:') === 0) {
        retina = true;
        data = data.replace(/^2x:/, '');
    }

    if (data.indexOf('legacy:') === 0) {
        legacy = true;
        data = data.replace(/^legacy:/, '');
    }

    var parsed;
    try {
        parsed = JSON.parse(data);
    } catch (e) {
        return callback('invalid geojson');
    }

    mapnikify(parsed, retina, function(err, xml) {
        if (err) return callback(err);
        this._xml = xml;
        this._size = retina && !legacy ? 512 : 256;
        this._bufferSize = retina ? 128 : 64;
        callback(null, this);
    }.bind(this));
}

/**
 * Gets a tile from this source.
 *
 * @param {number} z
 * @param {number} x
 * @param {number} y
 * @param {function} callback
 */
Source.prototype.getTile = function(z, x, y, callback) {
    var size = this._size;
    var map = new mapnik.Map(size, size);
    map.bufferSize = this._bufferSize;

    try {
        map.fromString(this._xml, {}, function(err) {
            if (err) return callback(err);
            map.extent = sm.bbox(x, y, z, false, '900913');
            map.render(new mapnik.Image(size, size), {}, onrender);
        });
    } catch (e) {
        callback(e);
    }

    function onrender(err, im) {
        if (err) return callback(err);
        im.encode('png8:m=h:z=1', function(err, res) {
            callback(err, res);
        });
    }
};

/**
 * Gets a grid from this source: this will always fail, because
 * this source does not provide grids.
 *
 * @param {number} z
 * @param {number} x
 * @param {number} y
 * @param {function} callback
 */
Source.prototype.getGrid = function(z, x, y, callback) {
    callback('This source does not provide grids');
};

/**
 * Gets info from this source: this will always fail, because
 * this source does not provide info.
 *
 * @param {function} callback
 */
Source.prototype.getInfo = function(callback) {
    callback('This source does not provide info');
};

/**
 * @param {object} tilelive
 */
Source.registerProtocols = function(tilelive) {
    tilelive.protocols['overlaydata:'] = Source;
};
